For #26644 - Apply the wallpaper provided text color to the Fenix logo

Mugurell opened the pull request:
The Fenix logo from homescreen was split in two separate images with the one
containing just text now being themable along with the private theme button.
They will be using the wallpaper text color when available.

The values for the text height and the top margin are taken from the Figma
designs - https://www.figma.com/file/d7l1iGabQYqnLcnQbHDGzM/Wallpaper-2.0-%2B-Onboarding-v106?node-id=7771%3A217341


Testing before and after the logo seem visually the same but it's not quite pixel
perfect. It may be because the Figma dimensions needed to be rounded.
Here I have applied a red tint just to show the theming working:

| | release or beta |
| --- | --- |
| current | <img src="https://user-images.githubusercontent.com/11428869/190399966-2c4a7104-9211-4f5f-9910-38efe51ef94b.png" /> |
| updated | <img src="https://user-images.githubusercontent.com/11428869/190402627-9f6f2396-1d81-487f-971c-cac3575b58e2.png" /> |
| updated clear | <img src="https://user-images.githubusercontent.com/11428869/190403204-6e6dffcc-c987-434a-8d02-9a975596b85c.png" /> |

| | nightly |
| --- | --- |
| current | <img src="https://user-images.githubusercontent.com/11428869/190401034-ab05e0c5-c1e6-4626-a933-a64a85e475d4.png" /> |
| updated | <img src="https://user-images.githubusercontent.com/11428869/190401210-f4544688-511e-4e5d-8ad2-767c3b256d87.png" /> |
| updated clear | <img src="https://user-images.githubusercontent.com/11428869/190401127-3e9fb417-0b7a-46cc-a8cd-5e8bf2b55d5b.png" /> |


| | debug |
| --- | --- |
| current | <img src="https://user-images.githubusercontent.com/11428869/190403802-ddf8f1ce-3d1f-42b8-831e-8c6d8ccc7d90.png" /> |
| updated | <img src="https://user-images.githubusercontent.com/11428869/190401589-a0fe293d-9579-4336-bdab-4e056a767190.png" /> |
| updated clear | <img src="https://user-images.githubusercontent.com/11428869/190401628-746231e2-4017-479e-9e6d-b45796848eb4.png" /> |






### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA
<!-- Before submitting the PR, please address each item -->
- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.








### GitHub Automation
Fixes #26644

MatthewTighe reviewed (commented):
Small suggestion, but this otherwise looks good to me. I would like to make sure design can review your concerns before landing though

MozillaNoah commented on app/src/nightly/res/values/dimens.xml:
```diff
@@ -0,0 +1,9 @@
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<resources>
+    <!-- Home Fragment -->
+    <dimen name="workmark_text_height">28dp</dimen>
+    <dimen name="workmark_text_margin_top">12dp</dimen>
```
```suggestion
    <dimen name="wordmark_text_height">28dp</dimen>
    <dimen name="wordmark_text_margin_top">12dp</dimen>
```

MozillaNoah commented:
I wonder if we could save on APK size by replacing the PNGs of the words with vectors. If they turn out to be the same for every version variant except for color, we could just have the one SVG and just re-color it.

Kate-Galetski commented:
Hi, 

1) It looks to be that the Firefox Preview is not aligned with the Firefox ball. Same goes with Firefox Browser. Nightly looks good.

2) Also could i please see the full homepage screenshot, where logo has the assigned colors: _Light / Text Primary_ or _Light text on color primary._

<img width="742" alt="Screen Shot 2022-09-15 at 11 15 16 AM" src="https://user-images.githubusercontent.com/87550321/190479564-d6b6326a-0a9d-417a-8e30-67870c2a9c4b.png">
<img width="781" alt="Screen Shot 2022-09-15 at 11 15 52 AM" src="https://user-images.githubusercontent.com/87550321/190479573-7fcb13d7-6231-4872-a788-b2d498c3e9f8.png">

Mugurell commented:
> It looks to be that the Firefox Preview is not aligned with the Firefox ball. Same goes with Firefox Browser. Nightly looks good.


If testing side by side and drawing straight lines to serve as guidance it seems like the proportions are met.
I also used the same ~[10px top padding](https://github.com/mozilla-mobile/fenix/blob/2d65d56bfbc14b76346fd0007285a31c60ea05eb/app/src/main/res/values/dimens.xml#L82) as seen on [Figma](https://www.figma.com/file/d7l1iGabQYqnLcnQbHDGzM/Wallpaper-2.0-%2B-Onboarding-v106).

(Images can be opened in a new tab to see them bigger)
![FirefoxSideBySide](https://user-images.githubusercontent.com/11428869/190599381-29eecef5-de19-4108-9a0d-82e8f61d6d2a.png)

![FenixSideBySide](https://user-images.githubusercontent.com/11428869/190601869-4cbaceb2-6ff5-400b-a1f4-ed666a8420aa.png)

Mugurell commented:
> Also could i please see the full homepage screenshot, where logo has the assigned colors: Light / Text Primary or Light text on color primary.

Posted videos on Jira - https://mozilla-hub.atlassian.net/browse/FNXV2-21392

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -968,11 +969,30 @@ class HomeFragment : Fragment() {
+     * Apply a color better contrasting with the current wallpaper to the Fenix logo and private mode switcher.
+     */
+    private fun applyWallpaperTextColor() {
+        if (shouldEnableWallpaper()) {
```
nit: This shouldn't strictly be necessary because it is checked in `applyWallpaper` as well but I'm fine with keep it if we want the extra assurance

Mugurell replied:
Will remove it.

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -968,11 +969,30 @@ class HomeFragment : Fragment() {
+     */
+    private fun applyWallpaperTextColor() {
+        if (shouldEnableWallpaper()) {
+            val tintColor = when (val color = requireContext().settings().currentWallpaperTextColor.toInt()) {
+                0 -> null // a null ColorStateList will clear the current tint
+                else -> ColorStateList.valueOf(color)
+            }
```
Interesting that `ColoStateList` requires an `Int` but other hex values expect `Long`s. I'm not sure there's a good way for us to avoid any errors in the conversion outside of relying on QA

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpapersUseCases.kt:
```diff
@@ -485,6 +485,7 @@ class WallpapersUseCases(
 
         private fun selectWallpaper(wallpaper: Wallpaper) {
             settings.currentWallpaperName = wallpaper.name
+            settings.currentWallpaperTextColor = wallpaper.textColor ?: 0
```
It would be good to verify this storage in a test IMO

Mugurell replied:
Added tests.

MatthewTighe reviewed (approved):
I see @Kate-Galetski left an affirmative comment in Jira, so I'm going to assume this is good from a UX perspective. I would prefer if my comment about a test was addressed, but this otherwise LGTM so r+ in case you want to land when you're online next

MatthewTighe reviewed (commented):
Ah, it looks like the private browsing color might be used in dark mode which is pretty illegible IMO

![dark-mode-text](https://user-images.githubusercontent.com/10427470/191128168-8c570411-5783-4fc7-9797-4b61666d5f59.png)

Mugurell commented:
Good catch with the logo color in dark theme, updated now to apply the same default tint as for the private theme when in dark mode:

https://user-images.githubusercontent.com/11428869/191248059-51854007-1d28-426d-bc65-4fe429dbb345.mov

Mugurell commented:
Some unit test are failing with seemingly no relation to these changes 🤔
https://firefoxci.taskcluster-artifacts.net/DzvUnQygTMWIknkdULRZrw/0/public/reports/test/testDebugUnitTest/index.html

Mugurell commented:
> Some unit test are failing with seemingly no relation to these changes 🤔 https://firefoxci.taskcluster-artifacts.net/DzvUnQygTMWIknkdULRZrw/0/public/reports/test/testDebugUnitTest/index.html

Failures seemed to be linked to the new tests I added in `HomeFragmentTest` - https://github.com/mozilla-mobile/fenix/blob/fc687e89a44c399f63ab5bfd4a27e9bd44c866c6/app/src/test/java/org/mozilla/fenix/home/HomeFragmentTest.kt.
Locally, running all the tests in that class alone works, running all tests with `./gradlew testDebugUnitTest` does indeed show the same failing tests as on TC. Not sure why they fail. To unblock merging I dropped these changes.

MatthewTighe commented:
@mergify backport releases_v106.0.0

MatthewTighe commented:
QA has verified so I created and merged a backport

Mugurell commented:
> QA has verified so I created and merged a backport

Think we also need the followup - https://github.com/mozilla-mobile/fenix/pull/27105, I'll create a backport PR for it and let for you to decide.
